Rocket.Chat/Managers/Requests/InfoRequestHandler.swift

```diff
@@ -0,0 +1,134 @@
+//  Copyright © 2017 Rocket.Chat. All rights reserved.
+//
+
+import Foundation
+import semver
+
+protocol InfoRequestHandlerDelegate: class {
+    var viewControllerToPresentAlerts: UIViewController? { get }
+
+    func urlNotValid()
+    func serverIsValid()
+    func serverChangedURL(_ newURL: String?)
+}
+
+class InfoRequestHandler: NSObject {
+
+    weak var delegate: InfoRequestHandlerDelegate?
+    var url: URL?
+    var validateServerVersion = true
+
+    func validate(with url: URL) {
+        API.shared.host = url
+        API.shared.fetch(InfoRequest(), sessionDelegate: self) { [weak self] result in
+            self?.validateServerResponse(result: result)
+        }
+    }
+
+    func alertInvalidURL() {
+        let alert = UIAlertController(
+            title: localized("alert.connection.invalid_url.title"),
+            message: localized("alert.connection.invalid_url.message"),
+            preferredStyle: .alert
+        )
+
+        alert.addAction(UIAlertAction(title: localized("global.ok"), style: .default, handler: nil))
+        delegate?.viewControllerToPresentAlerts?.present(alert, animated: true, completion: nil)
+    }
+
+    internal func validateServerResponse(result: InfoResult?) {
+        guard let version = result?.version else {
+            delegate?.urlNotValid()
+            return
+        }
+
+        if validateServerVersion {
+            if let minVersion = Bundle.main.object(forInfoDictionaryKey: "RC_MIN_SERVER_VERSION") as? String {
+                validateServerVersion(minVersion: minVersion, version: version)
+            }
+        }
+
+        delegate?.serverIsValid()
+    }
+
+    internal func validateServerVersion(minVersion: String, version: String) {
+        if Semver.lt(version, minVersion) {
+            let alert = UIAlertController(
+                title: localized("alert.connection.invalid_version.title"),
+                message: String(format: localized("alert.connection.invalid_version.message"), version, minVersion),
+                preferredStyle: .alert
+            )
+
+            alert.addAction(UIAlertAction(title: localized("global.ok"), style: .default, handler: nil))
+
+            if let controller = delegate?.viewControllerToPresentAlerts {
+                controller.present(alert, animated: true, completion: nil)
+            }
+        }
+    }
+
+}
+
+extension InfoRequestHandler: URLSessionTaskDelegate {
+
+    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
+        task.suspend()
+
+        if let location = response.allHeaderFields["Location"] as? String {
+            if let newURL = transformNewURL(location) {
+                handleRedirect(newURL)
+            }
+        }
+
+        completionHandler(nil)
+    }
+
+    func transformNewURL(_ newURL: String) -> URL? {
```

Comment:
We have URL.init(string:, scheme:) for this. Can we remove this and use that instead?

Reply:
Done, good finding!